File: ios/Classes/HelpshiftWrapperPlugin.swift
import Flutter
import UIKit
import HelpshiftX
import Foundation

public class HelpshiftWrapperPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "helpshift_wrapper", binaryMessenger: registrar.messenger())
    let instance = HelpshiftWrapperPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case Constant.methodConfigureSdk:
      do {
        self.configureHelpShiftSdk(call.arguments as! [String: Any])
        result("iOS Setup Completed")
      }
    case Constant.allConversation:
      do {
        let viewController1: UIViewController =
          (UIApplication.shared.delegate?.window??.rootViewController)!
        //          let tags = ["Hello", "iOS"]
        //          let config = ["tags":tags]
        let map = call.arguments as! [String: Any]
        let configMap = (map["configMap"])! as! [String: Any]
        self.showAllConversation(viewController1, configMap)
      }
    case Constant.allFAQs:
      do {
        let viewController1: UIViewController =
          (UIApplication.shared.delegate?.window??.rootViewController)!
        let map = call.arguments as! [String: Any]
        let configMap = (map["configMap"])! as! [String: Any]
        self.showAllFaqs(viewController1, configMap)
      }
    case Constant.methodSingleFAQsSection:
      do {
        let map = call.arguments as! [String: Any]
        let sectionId = (map["sectionId"] ?? "")! as! String

        let configMap = (map["configMap"])! as! [String: Any]
        self.showSingleFaqScreen(sectionId, configMap)
      }
    case Constant.methodSingleFAQDetail:
      do {
        let map = call.arguments as! [String: Any]
        let publishId = (map["publishId"])! as! String
        let configMap = (map["configMap"])! as! [String: Any]
        self.showSingleFAqDetail(publishId, configMap)
      }

    case Constant.methodLogin:
      do {
        let map = call.arguments as! [String: String]
        self.loginUser(map)
        result(true)
      }

    case Constant.methodLogout:
      do {
        self.logoutUser()
        result(true)
      }

    case Constant.methodSetLanguage:
      do {
        let map = call.arguments as! [String: Any]
        let language = (map["language"])! as! String
        self.setLanguage(language)
        result(true)
      }

    case Constant.methodHandleProactiveLink:
      do {
        let map = call.arguments as! [String: Any]
        let proactiveLink = (map["proactiveLink"])! as! String
        self.handleProactiveLink(proactiveLink)
        result(true)
      }

    case Constant.methodHandlePush:
      do {
        let map = call.arguments as! [String: Any]
        let data = (map["data"])! as! [String: String]
        self.handlePush(data)
        result(true)
      }

    case Constant.methodClearAnonymousUserOnLogin:
      do {
        let map = call.arguments as! [String: Any]
        let clearAnonymousUser = (map["clearAnonymousUser"])! as! Bool
        self.clearAnonymousUserOnLogin(clearAnonymousUser)
        result(true)
      }

    case Constant.methodRequestUnreadMessageCount:
      do {
        let map = call.arguments as! [String: Any]
        let shouldFetchFromServer = (map["shouldFetchFromServer"])! as! Bool
        self.requestUnreadMessageCount(shouldFetchFromServer)
        result(true)
      }

    default:
      do {

      }
      break

    }
    result("iOS " + UIDevice.current.systemVersion)
  }

  func configureHelpShiftSdk(_ map: [String: Any]) {
//    let apiKey = (map["helpShiftApiKey"])! as! String
    let domainName = (map["helpShiftDomain"])! as! String
    let appId = (map["helpShiftAppId"])! as! String

    let config = (map["configMap"])! as! [String: Any]

    Helpshift.install(withPlatformId: appId, domain: domainName, config: config)

    //        HelpshiftCore.initialize(with: HelpshiftSupport.sharedInstance())
    //        HelpshiftCore.install(forApiKey:apiKey, domainName: domainName, appID: appId)

  }

  func showAllConversation(_ type: UIViewController, _ configMap: [String: Any]) {

    Helpshift.showConversation(with: type, config: configMap)
    //  HelpshiftSupport.showConversation(type, config: configMap)
  }

  func showAllFaqs(_ type: UIViewController, _ configMap: [String: Any]) {
    Helpshift.showFAQs(with: type, config: configMap)
  }

  func showSingleFaqScreen(_ sectionId: String, _ configMap: [String: Any]) {

    let viewController: UIViewController =
      (UIApplication.shared.delegate?.window??.rootViewController)!
    Helpshift.showFAQSection(sectionId, with: viewController, config: configMap)
  }

  func showSingleFAqDetail(_ publishId: String, _ configMap: [String: Any]) {
    let viewController: UIViewController =
      (UIApplication.shared.delegate?.window??.rootViewController)!
    Helpshift.showSingleFAQ(publishId, with: viewController, config: configMap)
  }

  func loginUser(_ userDetail: [String: String]) {
    Helpshift.loginUser(userDetail)
  }

  func logoutUser() {
    Helpshift.logout()
  }

  func setLanguage(_ language: String) {
    Helpshift.setLanguage(language)
  }
    
  func handleProactiveLink(_ proactiveLink: String) {
    Helpshift.handleProactiveLink(proactiveLink)
  }

  func handlePush(_ data: [String: String]) {
      let viewController: UIViewController =
        (UIApplication.shared.delegate?.window??.rootViewController)!
    Helpshift.handleNotification(withUserInfoDictionary: data, isAppLaunch: true, with: viewController)
  }

  func clearAnonymousUserOnLogin(_ clearAnonymousUser: Bool) {
    Helpshift.clearAnonymousUser(onLogin: clearAnonymousUser)
  }

  func requestUnreadMessageCount(_ shouldFetchFromServer: Bool) {
    Helpshift.requestUnreadMessageCount(shouldFetchFromServer)
  }
}

File: android/src/main/kotlin/com/app/helpshift_wrapper/HelpshiftWrapperPlugin.kt
package com.app.helpshift_wrapper

import android.app.Activity
import com.helpshift.Helpshift
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** HelpshiftWrapperPlugin */
class HelpshiftWrapperPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "helpshift_wrapper")
        channel.setMethodCallHandler(this)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onMethodCall(call: MethodCall, result: Result) {

//        var status = await _channel.invokeListMethod("configure_sdk",
//        {'secret': helpShiftSecretKey, 'publishKey': helpShiftPublishableKey});
        when (call.method) {
            Constants.methodConfigureSdk -> {
                // configure and setup your sdk here
                val map = call.arguments as Map<String, String>
                configureHelpShiftSDK(map)
                result.success("SDK CONFIGURED")
            }

            Constants.allConversation -> {
                val map = call.arguments as Map<String, Any>
                val configMap = map["configMap"] as Map<String, Any>
                showAllConversation(configMap)
            }

            Constants.allFAQs -> {
                val map = call.arguments as Map<String, Any>
                val configMap = map["configMap"] as Map<String, Any>
                showAllFAQs(configMap)
            }

            Constants.methodSingleFAQsSection -> {
                val map = call.arguments as Map<String, Any>
                val sectionId = map["sectionId"] as String
                val configMap = map["configMap"] as Map<String, Any>
                openSingleFAQSectionScreen(sectionId, configMap)
            }

            Constants.methodSingleFAQDetail -> {
                val map = call.arguments as Map<String, String>
                val publishId = map["publishId"] as String
                val configMap = map["configMap"] as Map<String, Any>
                showSinglePublishedFAQ(publishId, configMap)
            }

            Constants.methodLogin -> {
                val map = call.arguments as Map<String, String>

                CoroutineScope(Dispatchers.Main).launch {
                    loginUser(map)
                    result.success(true)
                }
            }

            Constants.methodLogout -> {
                try {
                    logoutUser()
                    result.success(true)
                } catch (e: Exception) {
                    result.success(false)
                }
            }

            Constants.methodSetLanguage -> {
                val map = call.arguments as Map<String, Any>
                val language = map["language"] as String
                setLanguage(language)
                result.success(true)
            }

            Constants.methodHandleProactiveLink -> {
                val map = call.arguments as Map<String, Any>
                val proactiveLink = map["proactiveLink"] as String
                handleProactiveLink(proactiveLink)
                result.success(true)
            }

            Constants.methodHandlePush -> {
                val map = call.arguments as Map<String, Any>
                val data = map["data"] as Map<String, String>
                handlePush(data)
                result.success(true)
            }

            Constants.methodClearAnonymousUserOnLogin -> {
                val map = call.arguments as Map<String, Any>
                val clearAnonymousUser = map["clearAnonymousUser"] as Boolean
                clearAnonymousUserOnLogin(clearAnonymousUser)
                result.success(true)
            }

            Constants.methodRequestUnreadMessageCount -> {
                val map = call.arguments as Map<String, Any>
                val shouldFetchFromServer = map["shouldFetchFromServer"] as Boolean
                requestUnreadMessageCount(shouldFetchFromServer)
                result.success(true)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun configureHelpShiftSDK(args: Map<String, String>) {

        val configurations: HashMap<String, Any> = HashMap()
        configurations["enableInAppNotification"] = true
        configurations["enableLogging"] = true
        configurations["notificationChannelId"] = "HelpShift Notifications"
        configurations["notificationChannelId"] = "HelpShift Notifications"

        try {
            Helpshift.install(
                    activity!!.application,
                    args["helpShiftAppId"].toString(),
                    args["helpShiftDomain"].toString(),
                    configurations,
            )
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {

        }
    }

    private fun showAllConversation(configMap: Map<String, Any>) {
        Helpshift.showConversation(activity!!, configMap)
    }

    private fun showAllFAQs(configMap: Map<String, Any>) {
        Helpshift.showFAQs(activity!!, configMap)
    }

    private fun openSingleFAQSectionScreen(sectionId: String, configMap: Map<String, Any>) {
        Helpshift.showFAQSection(activity!!, sectionId, configMap)
    }

    private fun showSinglePublishedFAQ(publishId: String, configMap: Map<String, Any>) {
        Helpshift.showSingleFAQ(activity!!, publishId, configMap)
    }

    private fun loginUser(user: Map<String, String>) {
        Helpshift.login(user)
    }

    private fun logoutUser() {
        Helpshift.logout()
    }

    private fun setLanguage(language: String) {
        Helpshift.setLanguage(language)
    }

    private fun handleProactiveLink(proactiveLink: String) {
        Helpshift.handleProactiveLink(proactiveLink)
    }

    private fun handlePush(data: Map<String, String>) {
        Helpshift.handlePush(data)
    }

    private fun clearAnonymousUserOnLogin(clearAnonymousUser: Boolean) {
        Helpshift.clearAnonymousUserOnLogin(clearAnonymousUser)
    }

    private fun requestUnreadMessageCount(shouldFetchFromServer: Boolean) {
        Helpshift.requestUnreadMessageCount(shouldFetchFromServer)
    }
}
